src/target/riscv/riscv-013.c

```diff
@@ -279,16 +281,21 @@ dm013_info_t *get_dm(struct target *target)
 	return dm;
 }
 
-static uint32_t set_hartsel(uint32_t initial, uint32_t index)
+static uint32_t set_hartsel(uint32_t initial, int index)
 {
+	initial &= ~DM_DMCONTROL_HASEL;
 	initial &= ~DM_DMCONTROL_HARTSELLO;
 	initial &= ~DM_DMCONTROL_HARTSELHI;
 
-	uint32_t index_lo = index & ((1 << DM_DMCONTROL_HARTSELLO_LENGTH) - 1);
-	initial |= index_lo << DM_DMCONTROL_HARTSELLO_OFFSET;
-	uint32_t index_hi = index >> DM_DMCONTROL_HARTSELLO_LENGTH;
-	assert(index_hi < 1 << DM_DMCONTROL_HARTSELHI_LENGTH);
-	initial |= index_hi << DM_DMCONTROL_HARTSELHI_OFFSET;
+	if (index > 0) {
+		uint32_t index_lo = index & ((1 << DM_DMCONTROL_HARTSELLO_LENGTH) - 1);
+		initial |= index_lo << DM_DMCONTROL_HARTSELLO_OFFSET;
+		uint32_t index_hi = index >> DM_DMCONTROL_HARTSELLO_LENGTH;
+		assert(index_hi < 1 << DM_DMCONTROL_HARTSELHI_LENGTH);
+		initial |= index_hi << DM_DMCONTROL_HARTSELHI_OFFSET;
+	} else if (index < 0) {
+		initial |= DM_DMCONTROL_HASEL;
+	}
```

Comment:
Please, use an easier-to-read assertion.

I'd also recommend to use `index >= 0` in the first condition (even though in case of `index==0` the operations are redundant). But still, clarity of the code here has higher value than the few extra spent CPU cycles :)

Can we also use `set_field()` here instead of the bitmanip operators?

There is also `BIT_MASK` macro in helpers/bits.h that can be used here.

```suggestion
	const int MAX_HART_ID = (1 << (DM_DMCONTROL_HARTSELLO_LENGTH + DM_DMCONTROL_HARTSELHI_LENGTH)) - 1;
	/* -1 means multiple harts are selected */	
	assert(index == -1 || (index >= 0 && index <= MAX_HART_ID));
	
	/* Fill new values to the fields */
	if (index >= 0) {
		uint32_t index_lo = index & BIT_MASK(DM_DMCONTROL_HARTSELLO_LENGTH);
		uint32_t index_hi = index >> DM_DMCONTROL_HARTSELLO_LENGTH;				
		initial = set_field(initial, DM_DMCONTROL_HARTSELLO, index_lo);
		initial = set_field(initial, DM_DMCONTROL_HARTSELHI, index_hi);
	} else if (index < 0) {
		initial |= DM_DMCONTROL_HASEL;
	}
```

The `MAX_HART_ID` constant could also go to one of the common header files.

Reply:
Mostly done.

Reply:
Thank you.